src/TgSharp.Core/MTProto/Crypto/AesCtr.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TgSharp.Core.MTProto.Crypto
+{
+    /*
+     * AES-256-CTR Implementation 
+     * Original implementation by OpenSSL, ported to C#
+     */
+    public class AesCtr
+    {
+        // The input encrypted as though 128bit counter mode is being used.  The extra
+        // state information to record how much of the 128bit block we have used is
+        // contained in *num, and the encrypted counter is kept in ecount_buf.  Both
+        // *num and ecount_buf must be initialised with zeros before the first call to
+        // CRYPTO_ctr128_encrypt().
+        //
+        // This algorithm assumes that the counter is in the x lower bits of the IV
+        // (ivec), and that the application has full control over overflow and the rest
+        // of the IV.  This implementation takes NO responsibility for checking that
+        // the counter doesn't overflow into the rest of the IV when incremented.
+
+        public static void Ctr128Encrypt(byte[] input, byte[] key, ref byte[] ivec, ref byte[] ecount_buf, ref int num, byte[] output)
```

Comment:
better name: eCountBuffer, not ecount_buf, or even also unabbreviate 'e'

Reply:
fixed